homeassistant/components/conversation.py

```diff
@@ -84,49 +82,65 @@ def async_setup(hass, config):
 
         conf.extend(_create_matcher(utterance) for utterance in utterances)
 
-    @asyncio.coroutine
-    def process(service):
+    async def process(service):
         """Parse text into commands."""
         text = service.data[ATTR_TEXT]
-        yield from _process(hass, text)
+        try:
+            await _process(hass, text)
+        except intent.IntentHandleError as err:
+            _LOGGER.error('Error processing %s: %s', text, err)
 
     hass.services.async_register(
         DOMAIN, SERVICE_PROCESS, process, schema=SERVICE_PROCESS_SCHEMA)
 
     hass.http.register_view(ConversationProcessView)
 
-    async_register(hass, intent.INTENT_TURN_ON,
-                   ['Turn {name} on', 'Turn on {name}'])
-    async_register(hass, intent.INTENT_TURN_OFF,
-                   ['Turn {name} off', 'Turn off {name}'])
-    async_register(hass, intent.INTENT_TOGGLE,
-                   ['Toggle {name}', '{name} toggle'])
+    # We strip trailing 's' from name because our state matcher will fail
+    # if a letter is not there. By removing 's' we can match singular and
+    # plural names.
+
+    async_register(hass, intent.INTENT_TURN_ON, [
+        'Turn [the] [a] {name}[s] on',
+        'Turn on [the] [a] [an] {name}[s]',
+    ])
+    async_register(hass, intent.INTENT_TURN_OFF, [
+        'Turn [the] [a] [an] {name}[s] off',
+        'Turn off [the] [a] [an] {name}[s]',
+    ])
+    async_register(hass, intent.INTENT_TOGGLE, [
+        'Toggle [the] [a] [an] {name}[s]',
+        '[the] [a] [an] {name}[s] toggle',
+    ])
 
     return True
 
 
 def _create_matcher(utterance):
     """Create a regex that matches the utterance."""
-    parts = re.split(r'({\w+})', utterance)
+    parts = re.split(r'({\w+}|\[[\w\s]+\]\s*)', utterance)
     group_matcher = re.compile(r'{(\w+)}')
+    optional_matcher = re.compile(r'\[([\w\s]+)\]\s*')
 
     pattern = ['^']
-
     for part in parts:
-        match = group_matcher.match(part)
+        group_match = group_matcher.match(part)
+        optional_match = optional_matcher.match(part)
 
-        if match is None:
+        if group_match is None and optional_match is None:
             pattern.append(part)
             continue
 
-        pattern.append('(?P<{}>{})'.format(match.groups()[0], r'[\w ]+'))
+        if group_match is not None:
+            pattern.append(
+                r'(?P<{}>[\w ]+?)\s*'.format(group_match.groups()[0]))
```

Comment:
Any reason we're using `\w ` (notice whitespace) here and `\w\s` before?

(Also: I know we're only matching alphanumeric chars here, but I think it's best practice to escape all regexes nevertheless; just a comment)

Reply:
The parts regex splits the string into three types: "NO MATCH", "SLOT MATCH" and "OPTIONAL MATCH".

You are commenting on the GROUP MATCH line. That is split by `{\w+}`. It means that we require slot names to not contain spaces but that they are allowed to match names with spaces. For example, `{name}` will match `Kitchen Lights`

Reply:
I've added some comments to this code. Also replaced matching `\s` with `<space>` as that is really only what I care about.